docs/source/language/snql.rst

```diff
@@ -1,5 +1,190 @@
-==============
-Querying Snuba
-==============
+=======================
+The SnQL query language
+=======================
 
-Snuba is queried through the `snuba sdk <https://getsentry.github.io/snuba-sdk/>`_
+This document describes the Snuba Query Language (SnQL). For more details on
+how to actually send a query to Snuba see :doc:`/query/overview`.
+
+This is the query structure.::
+
+    MATCH simple | join | subquery
+    SELECT [expressions] | [aggregations BY expressions]
+    ARRAY JOIN [column]
+    WHERE condition [AND | OR]* condition*
+    HAVING condition [AND | OR]* condition*
+    ORDER BY expressions ASC|DESC
+    LIMIT expression BY n
+    LIMIT n
+    OFFSET n
+    GRANULARITY n
+    TOTALS boolean
+
+
+These queries are sent as strings to the ``/:dataset/snql`` endpoint encoded in a
+JSON body of the form below.::
+
+    {
+        "query": "<query>",
+        "dataset": "<dataset>",
+        "consistent": bool,
+        "turbo": bool,
+        "debug": bool,
+    }
+
+The dataset is implied through the url used for the query. All of the fields except
+for ``query`` are optional in the JSON body.
+
+MATCH
+=====
+
+Our data model is represented by a graph of entities. This clause identifies
+the pattern of the subgraphs we are querying on. There are three types of
+MATCH clause that are currently supported:
+
+**Simple:**
+
+``MATCH (<entity> [SAMPLE n])``
+
+This is equivalent to all of our current queries. This is querying data from
+a single entity (Errors, Transactions etc.) It is possible to add an optional
+sample to the query by adding it with the entity.
+
+**Subquery:**
+
+``MATCH { <query> }``
+
+Inside the curly braces can be another SnQL query in its entirety. Anything
+in the SELECT/BY clause of the subquery will be exposed in the outer query
+using the aliases specified.
+
+**Join:**
+
+``MATCH (<alias>.<entity> [SAMPLE n]) -[<join>]-> (<alias>.<entity> [SAMPLE n])``
+
+With JOINs every entity must have an alias, which is a unique string.
+Sampling can also be applied to any of the entities in the join. The
+``<join>`` is a string that is specified in the Entity in Snuba, and
+is a short hand for a set of join conditions. It's possible to have more
+than one join clause, separated by commas.
+
+SELECT .. BY
+============
+
+This clause specifies which results should be returned in the output.
+If there is an aggregation, then the ``BY`` clause must exist and everything
```

Comment:
This isn't technically true, since it's possible to have an aggregate with no grouping columns.

Reply:
fixed